Fixed #16010 -- Added Origin header checking to CSRF middleware. (#13829)
django/django: docs/releases/4.0.txt

Miscellaneous
-------------

* Values in the setting :setting:`CSRF_TRUSTED_ORIGINS` setting must include
  the scheme (``'http://'`` or ``'https:://'``) instead of only the hostname.

>   If a value starts with a dot, it should now also include an asterisk before
>   the dot. For example, ``'.example.com'`` becomes ``'https://*.example.com'``.

Reviewer: A little rearrangement to match tenses

```suggestion
  Values that previously started with a dot should now include an asterisk before
  the dot. For example, change ``'.example.com'`` to ``'https://*.example.com'``.
```

* As CSRF protection now consults the ``Origin`` header, you may need to set

>   :setting:`CSRF_TRUSTED_ORIGINS`, particularly if you set
>   :setting:`CSRF_COOKIE_DOMAIN` (or :setting:`SESSION_COOKIE_DOMAIN` if
>   :setting:`CSRF_USE_SESSIONS` is enabled) to a value that starts with a dot to
>   allow requests from subdomains.

Reviewer: A rearrangement pushing the setitng names to the end, does it work?

```suggestion
  :setting:`CSRF_TRUSTED_ORIGINS`, particularly if you allow requests from
  subdomains by setting the cookie domain to a value starting with a dot, in
  either :setting:`CSRF_COOKIE_DOMAIN` or :setting:`SESSION_COOKIE_DOMAIN`
  if :setting:`CSRF_USE_SESSIONS` is enabled).
```

in June 2021. Django 3.2 will be supported until April 2024. Django 4.0
officially supports Oracle 19c.

.. _csrf-trusted-origins-changes-4.0:

``CSRF_TRUSTED_ORIGINS`` changes
--------------------------------

Format change
~~~~~~~~~~~~~

Values in the :setting:`CSRF_TRUSTED_ORIGINS` setting must include the scheme

> (e.g. ``'http://'`` or ``'https:://'``) instead of only the hostname.

Reviewer 2: typo: `https:://`

Reviewer 3: Good catch :dart:

  to ``reorder_tests()``. It now accepts an iterable of tests rather than a
  test suite, and returns an iterator of tests.

* Values in the setting :setting:`CSRF_TRUSTED_ORIGINS` setting must include
  the scheme (``'http://'`` or ``'https:://'``) instead of only the hostname.
  Values that started with a dot, must now also include an asterisk before the
  dot. For example, change ``'.example.com'`` to ``'https://*.example.com'``.

* As CSRF protection now consults the ``Origin`` header, you may need to set
  :setting:`CSRF_TRUSTED_ORIGINS`, particularly if you allow requests from
  subdomains by setting :setting:`CSRF_COOKIE_DOMAIN` (or
  :setting:`SESSION_COOKIE_DOMAIN` if :setting:`CSRF_USE_SESSIONS` is enabled)
  to a value starting with a dot.

> 

Reviewer 4: This change will effect quite a lot of people (≈everyone?). I think this could be pulled up to it's own section in `Backwards incompatible changes in 4.0`, plus being linked to from the entry in `CSRF`. 

I know people will get the System Check on first run, but it's kind of a breaking change, so a bit more highlighting maybe?

Author: done

Reviewer 4: Super. Thanks @timgraham — that looks great. I will look at the code changes today. 👍
